### ISSUE TITLE
dask.bag.random.sample returns dask.bag.core.Item instead of dask.bag

### ISSUE THREAD

**Issue by shangw-nvidia** [dask/dask#6906]
**What happened**:

Say I have a bag A. What I want do is to pair a bag B, which is a smaller than A in terms of the number of elements, with a sampled version of A. For example:
A = [1 2 3 4 5 6 7]
B = ['a', 'b', 'c', 'd']
I want C = [('a', 3), ('b', 6), ('c', 5), ('d', 1)]

Looking at the documentation on [`dask.bag.random.sample`](https://docs.dask.org/en/latest/bag-api.html#dask.bag.random.sample), it looks like I can achieve this by:
```
import dask.bag as db
from dask.bag import random as db_random
from dask.distributed import Client
...
client = Client(n_workers=24, threads_per_worker=4)
...
A = db_random.choices(A, k=B.count())
C = db.zip(B, A).compute() 
```

However, it turned out that an `AttributeError` was raised by `db.zip`:
```
  File ".../lib/python3.7/site-packages/dask/bag/core.py", line 1913, in bag_zip
    assert all(bag.npartitions == npartitions for bag in bags)
  File ".../lib/python3.7/site-packages/dask/bag/core.py", line 1913, in <genexpr>
    assert all(bag.npartitions == npartitions for bag in bags)
AttributeError: 'Item' object has no attribute 'npartitions'
```

**What you expected to happen**:

`dask.bag.random.sample` would return `bag` instead of an `Item`.

**Minimal Complete Verifiable Example**:

Refers to the example above.

**Anything else we need to know?**:

Although I'm currently testing it in a single-workstation setting, I would eventually want the same code to be scaled on multi-nodes.

Please let me know, for the same purpose stated above, if there would be a better way to achieve the same thing.

**Environment**:

- Dask version: 2.30.0
- Python version: 3.7.8
- Operating System: Ubuntu 18.04.4 LTS
- Install method (conda, pip, source): conda

**Comment by mrocklin** [comment]
I agree that this looks like a bug and would be good to fix.  It looks like
we want to use the `out_type` parameter to the reduction method in the call
to dask/bag/random.py::_sample

**Comment by shangw-nvidia** [comment]
@mrocklin After I fixed `out_type` as you suggested: https://github.com/shangw-nvidia/dask/commit/b9f617055be2c1be5856e79c28ec8a24708b9b07 , I'm facing another problem when the input population to choices is a `dask.bag<flatten, npartitions=166>`:
```
distributed.worker - WARNING -  Compute Failed
Function:  execute_task
args:      ((<function empty_safe_apply at 0x7f5188af1cb0>, functools.partial(<function _sample_map_partitions at 0x7f5188af6050>, k=<dask.bag.core.Item object at 0x7f51397bdb10>, replace=True), (<cyfunction concat at 0x7f5251436bb0>, [['Keith Hamilton may refer to:'], ['"Falling in Love" is a song by English singer-songwriter Taio Cruz, released as the eighth overall single and second single from the second re-release of his second studio album, "Rokstarr".', 'The single was released exclusively in America and Spain, and was available via digital download from 15 April 2011.', 'The music video was uploaded to YouTube on 18 February 2011.', 'The music video was released to YouTube on 18 February 2011.', 'It features Cruz performing the single whilst sitting at a piano in the window of a first floor apartment.', 'He precedes to perform the song there until the second chorus, in which he leaves the apartment, and arrives in a private club where begins to perform on the stage.', 'The lights then fad
kwargs:    {}
Exception: TypeError("object of type 'itertools.chain' has no len()")

Traceback (most recent call last):
  File "online.py", line 70, in <module>
    main(parser.parse_args())
  File "online.py", line 58, in main
    print(samples.count().compute())
  File "/home/shangw/workspace/dask/dask/base.py", line 224, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/home/shangw/workspace/dask/dask/base.py", line 512, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/home/shangw/miniconda3/envs/rapids-0.17/lib/python3.7/site-packages/distributed/client.py", line 2725, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
  File "/home/shangw/miniconda3/envs/rapids-0.17/lib/python3.7/site-packages/distributed/client.py", line 1992, in gather
    asynchronous=asynchronous,
  File "/home/shangw/miniconda3/envs/rapids-0.17/lib/python3.7/site-packages/distributed/client.py", line 833, in sync
    self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
  File "/home/shangw/miniconda3/envs/rapids-0.17/lib/python3.7/site-packages/distributed/utils.py", line 340, in sync
    raise exc.with_traceback(tb)
  File "/home/shangw/miniconda3/envs/rapids-0.17/lib/python3.7/site-packages/distributed/utils.py", line 324, in f
    result[0] = yield future
  File "/home/shangw/miniconda3/envs/rapids-0.17/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/home/shangw/miniconda3/envs/rapids-0.17/lib/python3.7/site-packages/distributed/client.py", line 1851, in _gather
    raise exception.with_traceback(traceback)
  File "/home/shangw/workspace/dask/dask/bag/core.py", line 2407, in empty_safe_apply
    return func(part)
  File "/home/shangw/workspace/dask/dask/bag/random.py", line 85, in _sample_map_partitions
    lx = len(population)
TypeError: object of type 'itertools.chain' has no len()
```
In this case, the above example would be:
```
A = [[1], [2, 3, 4], [5, 6], [7]]
B = ['a', 'b', 'c', 'd']
```
and I still want `C = [('a', 3), ('b', 6), ('c', 5), ('d', 1)]`. I'm trying to achieve this by:
```
import dask.bag as db
from dask.bag import random as db_random
from dask.distributed import Client
...
client = Client(n_workers=24, threads_per_worker=4)
...
A = db_random.choices(A.flatten(), k=B.count())
C = db.zip(B, A).compute() 
```
Do I have to `.compute()` the `A.flatten()` before I pass it into `choices`?

**Comment by shangw-nvidia** [comment]
Btw, I can submit a PR from my fork for the `out_type` bug if no one has not got to that already.

**Comment by shangw-nvidia** [comment]
I found a way to work around it through converting `A` to `dask.DataFrame`, use `DataFrame`'s `sample`, the converting it back to `Bag`:

```
A = A.flatten().map(lambda v: {'v': v},).to_dataframe(meta={'v': int},).sample(frac=1.0).to_bag().flatten()
C = db.zip(B, A).compute()
```

However, I would imagine this approach is paying the overhead of converting back and forth between `Bag` and `DataFrame`.

**Comment by shangw-nvidia** [comment]
@mrocklin @jsignell Just pinging to see if there's any update on this issue?

Thanks!

**Comment by jsignell** [comment]
Please submit the `out_type` PR if you have it handy. For the second issue `TypeError: object of type 'itertools.chain' has no len()` I wasn't able to reproduce. With your change to `out_type` the following works as expected:

```python
import dask.bag as db
from dask.bag import random as db_random

A = db.from_sequence([1, 2, 3, 4, 5, 6, 7])
B = db.from_sequence(['a', 'b', 'c', 'd'])

a = db_random.choices(A, k=B.count().compute()).repartition(4)

C = db.zip(B, a).compute() 
```

If we wanted to support returning partitioned data from `db_random.choices` then we could add another kwarg to `choices` maybe `split_out`

**Comment by shangw-nvidia** [comment]
@jsignell In order to reproduce the `TypeError: object of type 'itertools.chain' has no len()` issue, `A` has to be a `flatten` bag. For example:
```
import dask.bag as db
from dask.bag import random as db_random

A = db.from_sequence([[1, 2], [3, 4, 5], [6], [7]])
B = db.from_sequence(['a', 'b', 'c', 'd'])

a = db_random.choices(A.flatten(), k=B.count().compute()).repartition(4)

C = db.zip(B, a).compute()
```

**Comment by shangw-nvidia** [comment]
Created PR #6941

**Comment by jsignell** [comment]
> A has to be a flatten bag. For example:

Ah sorry I missed that `A` had changed.

**Comment by shangw-nvidia** [comment]
Happy new year! Just checking back to see if there's any insight on this `flatten` bag problem?

**Comment by jsignell** [comment]
I think it'll be fixed by https://github.com/dask/dask/pull/7027 and I think that's a reasonable thing to do per-partition. I'll just add a test and get someone who knows more about bags to review.

**Comment by ncclementi** [comment]
@jsignell @shangw-nvidia Checking in here, maybe I'm missing something but it seems this issue got solved in #7027, should we close it?

**Comment by jsignell** [comment]
Yep! Closing
